realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -859,21 +861,270 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFind(
+
+    SharedGroup::Handover <Query> *handoverQueryPtr = reinterpret_cast<SharedGroup::Handover <Query> * > (queryPtr);
+    try {
+        std::unique_ptr <SharedGroup::Handover<Query>> handoverQuery(handoverQueryPtr);
+
+        SG(bgSharedGroupPtr)->end_read();
+
+        SharedGroup::VersionID currentVersion = SG(bgSharedGroupPtr)->get_version_of_current_transaction();
+        bool isDifferentVersions = (currentVersion != handoverQuery->version);
+        if (isDifferentVersions) {
+            SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
+        } else {
+            SG(bgSharedGroupPtr)->begin_read();
+        }
+
+        std::unique_ptr <Query> query = SG(bgSharedGroupPtr)->import_from_handover(
+                std::move(handoverQuery));
+        if (isDifferentVersions) {
+            LangBindHelper::advance_read(*SG(bgSharedGroupPtr), *CH(replicationPtr));
+        }
+
+        Table *table = query->get_table().get();
+
+        if (!QUERY_VALID(env, query.get())) {
+            return -1;
+        }
+
+        // It's valid to go 1 past the end index
+        if ((fromTableRow < 0) || (S(fromTableRow) > table->size())) {
+            // below check will fail with appropriate exception
+            (void) ROW_INDEX_VALID(env, table, fromTableRow);
+            return -1;
+        }
+
+        size_t r = query->find(S(fromTableRow));
+        if (r == not_found) {
+            return -1;
+        } else {
+            // handover the result
+            Row row = (*table)[r];
+            std::unique_ptr <SharedGroup::Handover<Row>> handover = SG(
+                    bgSharedGroupPtr)->export_for_handover(row);
+            return reinterpret_cast<jlong>(handover.release());
+        }
+
+    } CATCH_STD()
+    return -1;
+}
+
+
 JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAll(
     JNIEnv* env, jobject, jlong nativeQueryPtr, jlong start, jlong end, jlong limit)
 {
-    Query* pQuery = Q(nativeQueryPtr);
-    Table* pTable = pQuery->get_table().get();
-    if (!QUERY_VALID(env, pQuery) ||
-        !ROW_INDEXES_VALID(env, pTable, start, end, limit))
+    Query* query = Q(nativeQueryPtr);
+    Table* table = query->get_table().get();
+    if (!QUERY_VALID(env, query) ||
+        !ROW_INDEXES_VALID(env, table, start, end, limit))
         return -1;
     try {
-        TableView* pResultView = new TableView( pQuery->find_all(S(start), S(end), S(limit)) );
-        return reinterpret_cast<jlong>(pResultView);
+        TableView* tableView = new TableView( query->find_all(S(start), S(end), S(limit)) );
+        return reinterpret_cast<jlong>(tableView);
     } CATCH_STD()
     return -1;
 }
 
+JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAllWithHandover
+  (JNIEnv *env, jobject, jlong bgSharedGroupPtr, jlong replicationPtr, jlong queryPtr, jlong start, jlong end, jlong limit)
+  {
+      // cast the pointer
```

Comment:
We added them when we had some issues in generating C++ test cases that would show how the Java code was running. Having them makes it easier to see how core is really called. They were not added consistently. E.g. subtables are not used. And building the query itself is simple, but year it could use it too.